src/ripple/app/ledger/impl/LedgerCleaner.cpp

```diff
@@ -96,20 +101,24 @@ class LedgerCleanerImp
     //
     //--------------------------------------------------------------------------
 
-    void onPrepare ()
+    void onPrepare () override
     {
     }
 
-    void onStart ()
+    void onStart () override
     {
-        startThread();
+        thread_ = std::thread {&LedgerCleanerImp::run, this};
     }
 
-    void onStop ()
+    void onStop () override
     {
-        m_journal.info << "Stopping";
-        signalThreadShouldExit();
-        notify();
+        JLOG (j_.info) << "Stopping";
+        {
+            std::lock_guard<std::mutex> lock (mutex_);
+            state_ = State::shouldExit;
+            wakeup_.notify_one();
```

Comment:
The notify can be moved outside the lock.  This may have performance benefits as Scott D describes.  It is probably minor, but I haven't measured.  I see no problem with moving the signal outside of the lock in this case.